File: kangaroo_app_sdk_ios/ios/Classes/features/publicBusinessSettings/PublicBusinessSettingsHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class PublicBusinessSettingsHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_public_business_settings"

    var eventChannel: String = "customer_sdk/events/get_public_business_settings"

    func onMethodCall(call: FlutterMethodCall) -> Void? {
        PublicBusinessSettingsHandler.getPublicBusinessSettings(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getPublicBusinessSettings(call: FlutterMethodCall) {
        PublicBusinessSettingsApi().getPublicBusinessSettings()

        

        

        
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        PublicBusinessSettingsApi().publicBusinessSettingsState
            .serializePublicBusinessSettingsState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}

File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/publicBusinessSettings/PublicBusinessSettingsHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.publicBusinessSettings

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.publicBusinessSettings.PublicBusinessSettingsApi
import features.publicBusinessSettings.models.PublicBusinessSettings

import features.publicBusinessSettings.serializePublicBusinessSettingsState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class PublicBusinessSettingsHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_public_business_settings"

    override val eventChannel: String
        get() = "customer_sdk/events/get_public_business_settings"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return getPublicBusinessSettings(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun getPublicBusinessSettings(call: MethodCall): String? {
            val result = PublicBusinessSettingsApi().getPublicBusinessSettings().toJsonResult<PublicBusinessSettings>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        PublicBusinessSettingsApi().publicBusinessSettingsState.serializePublicBusinessSettingsState().observe { result ->
            pushSerializedResultToEventSink<PublicBusinessSettingsApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}
